src/test/pmreorder_xadd_range/pmreorder_xadd_range.c

```diff
@@ -0,0 +1,175 @@
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * pmreorder_xadd_range.c -- test for store reordering using
+ * pmemobj_tx_xadd_range with NO_DRAIN flag
+ *
+ * usage: pmreorder_xadd_range g|c file
+ *
+ * g - write data within a transaction
+ * c - check data consistency
+ *
+ */
+
+#include <string.h>
+#include <stddef.h>
+
+#include "unittest.h"
+#include "util.h"
+#include "valgrind_internal.h"
+#include "tx.h"
+
+#define DATA_SIZE	20
+#define OBJ_TYPE_NUM	2
+#define TEST_VALUE	4
+
+TOID_DECLARE(struct object, OBJ_TYPE_NUM);
+
+struct object {
+	char data[DATA_SIZE];
+};
+
+/*
+ * constructor -- persistent allocation constructor which sets values
+ * in data array to provided argument value
+ */
+static int
+constructor(PMEMobjpool *pop, void *ptr, void *arg)
+{
+	struct object *o = (struct object *)ptr;
+	for (int i = 0; i < DATA_SIZE; ++i) {
+		o->data[i] = *(char *)(arg);
+	}
+	pmemobj_persist(pop, o->data, DATA_SIZE * sizeof(o->data[0]));
+	return 0;
+}
+
+/*
+ * write data -- write data to pool in transaction, using
+ * POBJ_XADD_NO_DRAIN flag in order to add multiple ranges with only one fence
+ * at the end
+ */
+static void
+write_data(PMEMobjpool *pop)
+{
+	TOID(struct object) obj;
+	int value = TEST_VALUE;
+	POBJ_ALLOC(pop, &obj, struct object, sizeof(struct object),
+				constructor, &value);
+
+	char tmp[DATA_SIZE];
+	memcpy(tmp, D_RO(obj)->data, DATA_SIZE);
+
+	TX_BEGIN(pop) {
+
+		VALGRIND_EMIT_LOG("TX_ADD.BEGIN");
+
+		TX_XADD_FIELD(obj, data[1], POBJ_XADD_NO_DRAIN);
+		D_RW(obj)->data[1] = 1;
```

Comment:
You can't modify data if you used NO_DRAIN flag and didn't drain before modification!

If pmemcheck doesn't detect this bug, then you need to investigate that and possibly adapt pmemcheck.